Issues running tests with Rmd notebooks

Hello,

I've gotten through the full workflow with the ipynb demo files in the tutorial documation, but I want to use otter to grade Rmd notebooks.

I have the following setup so far (all files can be found on my [github](https://github.com/hgibling/otter-test)):
```
cd ~/Documents/otter-test

.
├── TestAssignment.Rmd
├── dist
│   ├── autograder
│   │   ├── TestAssignment.Rmd
│   │   └── tests
│   │       ├── Question1.R
│   │       └── Question2.R
│   └── student
│       ├── TestAssignment.Rmd
│       └── tests
│           ├── Question1.R
│           └── Question2.R
└── submissions
    ├── Student1-Correct-TestAssignment.Rmd
    └── Student2-Wrong-TestAssignment.Rmd

```

Question1 has one open and one hidden test, and Question2 has 2 hidden tests. The student distribution notebook looks fine to me, and I copied them and put in answers to make the two submission files. I've run into three issues:

1. In R, the relative path to `tests` does not work in the student notebook, I need the full path (not the end of the world; the `dist/autograder` notebook can use relative paths fine):
```
> setwd("~/Documents/otter-test/dist/student/")
> 
> library(testthat)
> library(ottr)
>
> ice_cream <- "vanilla" # YOUR CODE HERE
> . = ottr::check("tests/Question1.R")
cannot open file 'tests/Question1.R': No such file or directoryError in file(filename, "r") : cannot open the connection
> getwd()
[1] "/Users/hgibling/Documents/otter-test/dist/student"
> . = ottr::check("~/Documents/otter-test/dist/student/tests/Question1.R")
All tests passed!
```

2. While the tests work appropriately in the original raw notebook, in the student notebooks (and in the `dist/autograder` notebook) the tests pass when the answer is blatantly wrong:
```
> getwd()
[1] "/Users/hgibling/Documents/otter-test/dist/student"
> ice_cream <- 13 # YOUR CODE HERE
> . = ottr::check("~/Documents/otter-test/dist/student/tests/Question1.R")
All tests passed!

###

> setwd("~/Documents/otter-test/dist/autograder/")
> ice_cream <- "chocolate" #SOLUTION
> . = ottr::check("tests/Question1.R")
All tests passed!
> ice_cream <- 13 #SOLUTION
> . = ottr::check("tests/Question1.R")
All tests passed!
```

3. I imagine the previous issue might have something to do with not being able to run `otter check` in terminal. Since I'm pretending to be a student, I am supplying the path to the student tests created after `otter assign` in `dist`:
```
cd ~/Documents/otter-test/submissions
otter check Student1-Correct-TestAssignment.Rmd -t ~/Documents/otter-test/dist/student/tests -q Question1 

Traceback (most recent call last):
  File "/Users/hgibling/miniconda3/bin/otter", line 8, in <module>
    sys.exit(cli())
  File "/Users/hgibling/miniconda3/lib/python3.7/site-packages/click/core.py", line 1137, in __call__
    return self.main(*args, **kwargs)
  File "/Users/hgibling/miniconda3/lib/python3.7/site-packages/click/core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "/Users/hgibling/miniconda3/lib/python3.7/site-packages/click/core.py", line 1668, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/hgibling/miniconda3/lib/python3.7/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/hgibling/miniconda3/lib/python3.7/site-packages/click/core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "/Users/hgibling/miniconda3/lib/python3.7/site-packages/otter/cli.py", line 55, in check_cli
    return check(*args, **kwargs)
  File "/Users/hgibling/miniconda3/lib/python3.7/site-packages/otter/check/__init__.py", line 77, in main
    raise e
  File "/Users/hgibling/miniconda3/lib/python3.7/site-packages/otter/check/__init__.py", line 50, in main
    assert os.path.isfile(test_path), "Test {} does not exist".format(question)
AssertionError: Test Question1.R does not exist
```

I'm not sure what's happening. Any ideas?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,7 @@
 **Unreleased:**
 
 * Fixed [#389](https://github.com/ucbds-infra/otter-grader/issues/389) by closing file handles in Otter Grade before removing them
+* Fixed R image builds by setting Conda `channel_priority` to `strict` in R `setup.sh` files per [#386](https://github.com/ucbds-infra/otter-grader/issues/386)
 
 **v3.1.2:**
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,9 @@
 # Changelog
 
+**Unreleased:**
+
+* Fixed [#389](https://github.com/ucbds-infra/otter-grader/issues/389) by closing file handles in Otter Grade before removing them
+
 **v3.1.2:**
 
 * Specify UTF-8 encoding in all `open` calls used for reading JSON for Windows compatibility per [#380](https://github.com/ucbds-infra/otter-grader/issues/380)

--- a/Dockerfile
+++ b/Dockerfile
@@ -61,4 +61,4 @@ RUN mkdir /autograder
 ADD requirements.txt /tmp/requirements.txt
 RUN pip install -r /tmp/requirements.txt
 
-RUN pip install otter-grader==3.1.2
+RUN pip install git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb

--- a/Dockerfile
+++ b/Dockerfile
@@ -61,4 +61,4 @@ RUN mkdir /autograder
 ADD requirements.txt /tmp/requirements.txt
 RUN pip install -r /tmp/requirements.txt
 
-RUN pip install git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb
+RUN pip install git+https://github.com/ucbds-infra/otter-grader.git@6bc9682296fce1667430b5286ebcb285cc2668de

--- a/otter/generate/templates/python/requirements.txt
+++ b/otter/generate/templates/python/requirements.txt
@@ -12,6 +12,6 @@ nbconvert
 nbformat
 dill
 numpy
-git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb
+git+https://github.com/ucbds-infra/otter-grader.git@6bc9682296fce1667430b5286ebcb285cc2668de
 {% endif %}{% if other_requirements %}
 {{ other_requirements }}{% endif %}

--- a/otter/generate/templates/python/requirements.txt
+++ b/otter/generate/templates/python/requirements.txt
@@ -12,6 +12,6 @@ nbconvert
 nbformat
 dill
 numpy
-otter-grader==3.1.2
+git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb
 {% endif %}{% if other_requirements %}
 {{ other_requirements }}{% endif %}

--- a/otter/generate/templates/r/requirements.txt
+++ b/otter/generate/templates/r/requirements.txt
@@ -13,4 +13,4 @@ nbformat
 dill
 rpy2
 numpy
-git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb
+git+https://github.com/ucbds-infra/otter-grader.git@6bc9682296fce1667430b5286ebcb285cc2668de

--- a/otter/generate/templates/r/requirements.txt
+++ b/otter/generate/templates/r/requirements.txt
@@ -13,4 +13,4 @@ nbformat
 dill
 rpy2
 numpy
-otter-grader==3.1.2
+git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb

--- a/otter/generate/templates/r/setup.sh
+++ b/otter/generate/templates/r/setup.sh
@@ -33,6 +33,7 @@ if [ "${BASE_IMAGE}" != "ucbdsinfra/otter-grader" ]; then
 fi
 
 # install dependencies with conda
+conda config --set channel_priority strict
 conda env create -f {{ autograder_dir }}/source/environment.yml
 conda run -n {{ otter_env_name }} Rscript {{ autograder_dir }}/source/requirements.r
 

--- a/otter/grade/containers.py
+++ b/otter/grade/containers.py
@@ -173,8 +173,9 @@ def grade_assignments(submission_path, image="ucbdsinfra/otter-grader", verbose=
             timer = threading.Timer(timeout, kill_container)
             timer.start()
 
+        container_id = container.id[:12]
         if verbose:
-            print(f"Grading {submission_path} in container {container.id}...")
+            print(f"Grading {submission_path} in container {container_id}...")
 
         exit = docker.container.wait(container)
 

--- a/otter/grade/containers.py
+++ b/otter/grade/containers.py
@@ -39,10 +39,9 @@ def build_image(zip_path, base_image, tag):
     return image
 
 
-def launch_grade(zip_path, notebooks_dir, verbose=False, num_containers=None,
-                 scripts=False, no_kill=False, output_path="./", debug=False, zips=False,
-                 image="ucbdsinfra/otter-grader",
-                 pdfs=False, timeout=None, network=True):
+def launch_grade(zip_path, notebooks_dir, verbose=False, num_containers=None, scripts=False, 
+                 no_kill=False, output_path="./", debug=False, zips=False,
+                 image="ucbdsinfra/otter-grader", pdfs=False, timeout=None, network=True):
     """
     Grades notebooks in parallel Docker containers
 
@@ -91,7 +90,6 @@ def launch_grade(zip_path, notebooks_dir, verbose=False, num_containers=None,
                 grade_assignments,
                 submission_path=nb_path,
                 verbose=verbose,
-                # TODO:check if path is not default for generate hash
                 image=img,
                 scripts=scripts,
                 no_kill=no_kill,
@@ -191,7 +189,7 @@ def grade_assignments(submission_path, image="ucbdsinfra/otter-grader", verbose=
         if exit != 0:
             raise Exception(f"Executing '{submission_path}' in docker container failed! Exit code: {exit}")
 
-        with open(results_file, "rb") as f:
+        with open(results_path, "rb") as f:
             scores = pickle.load(f)
 
         scores = scores.to_dict()

--- a/otter/grade/containers.py
+++ b/otter/grade/containers.py
@@ -138,13 +138,13 @@ def grade_assignments(submission_path, image="ucbdsinfra/otter-grader", verbose=
         ``pandas.core.frame.DataFrame``: A dataframe of file to grades information
     """
 
-    _, temp_subm_path = tempfile.mkstemp()
+    temp_subm_file, temp_subm_path = tempfile.mkstemp()
     shutil.copyfile(submission_path, temp_subm_path)
 
     results_file, results_path = tempfile.mkstemp(suffix=".pkl")
     pdf_path = None
     if pdfs:
-        _, pdf_path = tempfile.mkstemp(suffix=".pdf")
+        pdf_file, pdf_path = tempfile.mkstemp(suffix=".pdf")
 
     try:
         nb_basename = os.path.basename(submission_path)
@@ -206,9 +206,12 @@ def grade_assignments(submission_path, image="ucbdsinfra/otter-grader", verbose=
             shutil.copy(pdf_path, local_pdf_path)
 
     finally:
+        os.close(results_file)
         os.remove(results_path)
+        os.close(temp_subm_file)
         os.remove(temp_subm_path)
         if pdfs:
+            os.close(pdf_file)
             os.remove(pdf_path)
 
     return df

--- a/test/test-assign/gs-autograder-correct/requirements.txt
+++ b/test/test-assign/gs-autograder-correct/requirements.txt
@@ -12,4 +12,4 @@ nbconvert
 nbformat
 dill
 numpy
-git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb
+git+https://github.com/ucbds-infra/otter-grader.git@6bc9682296fce1667430b5286ebcb285cc2668de

--- a/test/test-assign/gs-autograder-correct/requirements.txt
+++ b/test/test-assign/gs-autograder-correct/requirements.txt
@@ -12,4 +12,4 @@ nbconvert
 nbformat
 dill
 numpy
-otter-grader==3.1.2
+git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb

--- a/test/test-assign/rmd-autograder-correct/requirements.txt
+++ b/test/test-assign/rmd-autograder-correct/requirements.txt
@@ -13,4 +13,4 @@ nbformat
 dill
 rpy2
 numpy
-git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb
+git+https://github.com/ucbds-infra/otter-grader.git@6bc9682296fce1667430b5286ebcb285cc2668de

--- a/test/test-assign/rmd-autograder-correct/requirements.txt
+++ b/test/test-assign/rmd-autograder-correct/requirements.txt
@@ -13,4 +13,4 @@ nbformat
 dill
 rpy2
 numpy
-otter-grader==3.1.2
+git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb

--- a/test/test-assign/rmd-autograder-correct/setup.sh
+++ b/test/test-assign/rmd-autograder-correct/setup.sh
@@ -33,6 +33,7 @@ if [ "${BASE_IMAGE}" != "ucbdsinfra/otter-grader" ]; then
 fi
 
 # install dependencies with conda
+conda config --set channel_priority strict
 conda env create -f /autograder/source/environment.yml
 conda run -n otter-env Rscript /autograder/source/requirements.r
 

--- a/test/test-run/autograder/source/requirements.txt
+++ b/test/test-run/autograder/source/requirements.txt
@@ -12,6 +12,6 @@ nbconvert
 nbformat
 dill
 numpy
-git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb
+git+https://github.com/ucbds-infra/otter-grader.git@6bc9682296fce1667430b5286ebcb285cc2668de
 
 tqdm

--- a/test/test-run/autograder/source/requirements.txt
+++ b/test/test-run/autograder/source/requirements.txt
@@ -12,6 +12,6 @@ nbconvert
 nbformat
 dill
 numpy
-otter-grader==3.1.2
+git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb
 
 tqdm

--- a/test/test_generate/test-autograder/autograder-correct/requirements.txt
+++ b/test/test_generate/test-autograder/autograder-correct/requirements.txt
@@ -12,6 +12,6 @@ nbconvert
 nbformat
 dill
 numpy
-git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb
+git+https://github.com/ucbds-infra/otter-grader.git@6bc9682296fce1667430b5286ebcb285cc2668de
 
 tqdm

--- a/test/test_generate/test-autograder/autograder-correct/requirements.txt
+++ b/test/test_generate/test-autograder/autograder-correct/requirements.txt
@@ -12,6 +12,6 @@ nbconvert
 nbformat
 dill
 numpy
-otter-grader==3.1.2
+git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb
 
 tqdm

--- a/test/test_generate/test-autograder/autograder-custom-env/requirements.txt
+++ b/test/test_generate/test-autograder/autograder-custom-env/requirements.txt
@@ -12,6 +12,6 @@ nbconvert
 nbformat
 dill
 numpy
-git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb
+git+https://github.com/ucbds-infra/otter-grader.git@6bc9682296fce1667430b5286ebcb285cc2668de
 
 tqdm

--- a/test/test_generate/test-autograder/autograder-custom-env/requirements.txt
+++ b/test/test_generate/test-autograder/autograder-custom-env/requirements.txt
@@ -12,6 +12,6 @@ nbconvert
 nbformat
 dill
 numpy
-otter-grader==3.1.2
+git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb
 
 tqdm

--- a/test/test_generate/test-autograder/autograder-token-correct/requirements.txt
+++ b/test/test_generate/test-autograder/autograder-token-correct/requirements.txt
@@ -12,6 +12,6 @@ nbconvert
 nbformat
 dill
 numpy
-git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb
+git+https://github.com/ucbds-infra/otter-grader.git@6bc9682296fce1667430b5286ebcb285cc2668de
 
 tqdm

--- a/test/test_generate/test-autograder/autograder-token-correct/requirements.txt
+++ b/test/test_generate/test-autograder/autograder-token-correct/requirements.txt
@@ -12,6 +12,6 @@ nbconvert
 nbformat
 dill
 numpy
-otter-grader==3.1.2
+git+https://github.com/ucbds-infra/otter-grader.git@6ee10a36b4267f924cceaf21f684f44fc5ab76bb
 
 tqdm

